pkg/oci/repository.go

```diff
@@ -0,0 +1,271 @@
+// The fetched blob is cached in the repository.
+func (r *Repository) FetchBlobFrom(s string, verify bool) (v1.Layer, error) {
+	return r.fetchBlobFrom(s, verify)
+}
+
+// FetchBlobFromUncompressed fetches a blob from the given remote. It accepts a string
+// and a boolean to indicate whether the blob should be verified. In order to verify the blob,
+// the compressed blob is read and the digest is verified against the digest in the string.
+// The uncompressed blob is returned.
+// The fetched blob is cached in the repository.
+func (r *Repository) FetchBlobFromUncompressed(s string, verify bool) (io.ReadCloser, error) {
+	l, err := r.fetchBlobFrom(s, verify)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch layer: %v", err)
+	}
+
+	// get uncompressed layer
+	return l.Uncompressed()
+}
+
+func (r *Repository) fetchBlobFrom(s string, verify bool) (v1.Layer, error) {
+	ref, err := name.NewDigest(s)
+	if err != nil {
+		return nil, fmt.Errorf("%s: blob s must be of the form <name@digest>", ref)
+	}
+	layer, err := remote.Layer(ref, r.remoteOpts...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch layer: %v", err)
+	}
+	if verify {
+		verified, err := verifyBlob(ref, layer)
+		if err != nil {
+			return nil, fmt.Errorf("failed to verify layer: %v", err)
+		}
+		if !verified {
+			return nil, fmt.Errorf("failed to verify layer")
+		}
+	}
+	// cache the fecched blob in the repository
+	err = r.PushBlob(layer)
+	if err != nil {
+		return nil, fmt.Errorf("failed to cache layer: %v", err)
+	}
+	return layer, nil
+}
+
+// FetchBlob fetches a blob from the repository.
+func (r *Repository) FetchBlob(digest string) (v1.Layer, error) {
+	ref, err := name.NewDigest(fmt.Sprintf("%s@%s", r.Repository, digest))
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse digest %q: %w", digest, err)
+	}
+	return remote.Layer(ref, r.remoteOpts...)
+}
+
+// FetchBlobUncompressed fetches a blob from the given remote.
+func (r *Repository) FetchBlobUncompressed(digest string) (io.ReadCloser, error) {
+	l, err := r.FetchBlob(digest)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch layer: %v", err)
+	}
+
+	// get uncompressed layer
+	return l.Uncompressed()
+}
+
+// PushBlob pushes a blob to the repository. It accepts a v1.Layer interface.
+func (r *Repository) PushBlob(layer v1.Layer) error {
+	return remote.WriteLayer(r.Repository, layer, r.remoteOpts...)
+}
+
+// CopyBlob copies a blob from the given remote to the location specified by the
+// repository. It accepts a s as a string and a boolean to indicate whether the blob should
+// be verified. In order to verify the blob, the compressed blob is read and the digest is verified
+// against the digest in the s.
+func (r *Repository) CopyBlob(s string, verify bool) error {
+	// fetch blob content
+	l, err := r.FetchBlobFrom(s, verify)
+	if err != nil {
+		return fmt.Errorf("failed to fetch layer: %v", err)
+	}
+	// push blob content
+	if err := r.PushBlob(l); err != nil {
```

Comment:
This is interesting. Why would we copy from our repository to our repository? For example, from a Snapshot to another Snapshot? Like, after configuration?

Reply:
We don't need this receiver function anymore. I will remove it.